pkg/scheduler/plugins/gang/gang.go

```diff
@@ -0,0 +1,428 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gang
+
+import (
+	"strconv"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/scheduler-plugins/pkg/apis/scheduling/v1alpha1"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+
+	schedulingconfig "github.com/koordinator-sh/koordinator/apis/scheduling/config"
+)
+
+var (
+	timeNowFn = time.Now
+)
+
+const (
+	GangFromPodGroupCrd   string = "GangFromPodGroupCrd"
+	GangFromPodAnnotation string = "GangFromPodAnnotation"
+)
+
+// Gang  basic gang info recorded in gangCache:
+type Gang struct {
+	Name       string
+	WaitTime   time.Duration
+	CreateTime time.Time
+
+	// time that first pod wait in Permit stage
+	TimeoutStartTime time.Time
+	// strict-mode or non-strict-mode
+	Mode              string
+	MinRequiredNumber int
+	TotalChildrenNum  int
+	GangGroup         []string
+	Children          map[string]*v1.Pod
+	// pods that have already assumed(waiting in Permit stage)
+	WaitingForBindChildren map[string]*v1.Pod
+	// pods that have already bound
+	BoundChildren map[string]*v1.Pod
+	// if assumed  pods number has reached to MinRequiredNumber
+	ResourceSatisfied bool
+
+	// if the gang should be passed at PreFilter stage(Strict-Mode)
+	ScheduleCycleValid bool
```

Comment:
please give some instruction for the following three members, ScheduleCycleValid, ScheduleCycle and ChildrenScheduleRoundMap.

Reply:
done